[{"attachment_id": null, "tags": [], "creator": "wr0ngway@yahoo.com", "text": "If I create a parser (SAX or DOM) and then try an parse an InputSource I create \ndirectly from a FileReader, I get a NullPointerException if the xml contained \nby the reader has any external parsed entities defined in it. If the xml has no \nexternal entities (internal ok), the parse succeeds, and if I parse using a \nString systemId instead of a InputSource, the parse also succeeds.  If there is \nsome precondition to using InputSources (like setting the systemId...?) then a \nmore helpful error message would be a good idea.\n\ne.g.\n\nfoo.xml:\n<?xml version=\"1.0\"?>\n\n<!DOCTYPE foo [\n   <!ENTITY baz SYSTEM \"baz.xml\">\n]>\n\n<foo/>\n\nbaz.xml:\n<baz/>\n\nfoo.java:\nimport org.xml.sax.*;\nimport org.xml.sax.helpers.*;\nimport java.io.*;\n\npublic class foo\n{\n\tpublic static void main (String[] args) \n\t{\n\t\ttry\n\t\t{\n\t\t\tString filename = args[0];\n\t\t\tXMLReader reader = XMLReaderFactory.createXMLReader\n(\"org.apache.xerces.parsers.SAXParser\");\n\t\t\tInputSource src = new InputSource(new BufferedReader\n(new FileReader(filename)));\n\t\t\t// reader.parse(filename); //this works\n\t\t\treader.parse(src); // This NPE's\n\t\t}\n\t\tcatch (Exception ex)\n\t\t{\n\t\t\tSystem.out.println(\"Unexpected exception: \" + ex);\n\t\t\tex.printStackTrace();\n\t\t}\n\t}\n\t\n}\n\njavac -classpath .;xerces.jar foo.java\njava -classpath .;xerces.jar foo foo.xml", "count": 0, "id": 6323, "time": "2001-10-03T11:06:28Z", "bug_id": 3945, "creation_time": "2001-10-03T11:06:28Z", "is_private": false}, {"count": 1, "text": "I have also encountered this problem. I recompiled Xerces to print stack traces \nand I get the following stack trace for the NPE:\n\njava.lang.NullPointerException\n\tat org.apache.xerces.utils.StringPool.addSymbol(StringPool.java:347)\n\tat org.apache.xerces.readers.DefaultEntityHandler.addExternalEntityDecl\n(DefaultEntityHandler.java:226)\n\tat org.apache.xerces.framework.XMLDTDScanner.scanEntityDecl\n(XMLDTDScanner.java:2800)\n\tat org.apache.xerces.framework.XMLDTDScanner.scanDecls\n(XMLDTDScanner.java:1450)\n\tat org.apache.xerces.framework.XMLDTDScanner.scanDoctypeDecl\n(XMLDTDScanner.java:1123)\n\tat org.apache.xerces.framework.XMLDocumentScanner.scanDoctypeDecl\n(XMLDocumentScanner.java:2145)\n\tat org.apache.xerces.framework.XMLDocumentScanner.access$0\n(XMLDocumentScanner.java:2100)\n\tat \norg.apache.xerces.framework.XMLDocumentScanner$PrologDispatcher.dispatch\n(XMLDocumentScanner.java:831)\n\tat org.apache.xerces.framework.XMLDocumentScanner.parseSome\n(XMLDocumentScanner.java:381)\n\tat org.apache.xerces.framework.XMLParser.parse(XMLParser.java:1081)\n\nThe problem seems to be that in DefaultEntityHandler.addExternalEntityDecl(), \nfSystemId is null. So when it tries to add this null String to the StringPool, \nthe StringPool throws an NPE.", "bug_id": 3945, "is_private": false, "id": 7265, "time": "2001-10-29T09:13:44Z", "creator": "gordon.tyler@sitraka.com", "creation_time": "2001-10-29T09:13:44Z", "tags": [], "attachment_id": null}, {"attachment_id": null, "tags": [], "creator": "gordon.tyler@sitraka.com", "is_private": false, "count": 2, "id": 7267, "time": "2001-10-29T10:06:35Z", "bug_id": 3945, "creation_time": "2001-10-29T10:06:35Z", "text": "After having debugged Xerces, I think I've figured what causes the problem.\n\nWhen an InputSource is created from a Reader, it's systemId is null. When a \ndocument is parsed, during the setup in XMLParser, \nDefaultEntityHandler.startReadingFromDocument() is called. It initialises it's \nfSystemId field from InputSource.getSystemId() and thus it is initialised to \nnull. So when DefaultEntityHandler.addExternalEntityDecl() is called later, it \npasses the null fSystemId to StringPool.addSymbol() and causes the NPE.\n\nIf the InputSource is created using a filename or created using a Reader and a \nsystemId, the problem does not occur.\n\nOne possible solution would be to make Xerces report an error if an external \nentity is encountered when a base systemId is not available (i.e. null)."}, {"count": 3, "tags": [], "bug_id": 3945, "attachment_id": null, "text": "This is a duplicate of bug 4026.  As Glenn Marcy has pointed out elsewhere, the \nInputSource object needs to have its System ID set in order to always correctly \nresolve references relative to it - e.g., the relative reference to \"baz.xml\" \nin the entity declaration inside foo.xml.\n\nYou can do that explicitly using the InputSource.setSystemId(String) method.\n\nIn any event, Xerces shouldn't be throwing a NullPointerException, so there is \nstill a bug there.\n\n\n*** This bug has been marked as a duplicate of 4026 ***", "id": 7272, "time": "2001-10-29T10:41:13Z", "creator": "zongaro@ca.ibm.com", "creation_time": "2001-10-29T10:41:13Z", "is_private": false}]